src/primitive/cuboid.rs

```diff
@@ -0,0 +1,222 @@
+use cgmath::{BaseFloat, Point3, Vector3};
+use cgmath::prelude::*;
+
+use {Aabb3, Ray3};
+use prelude::*;
+
+/// Cuboid primitive.
+///
+/// Have a cached set of corner points to speed up computation.
+#[derive(Debug, Clone, PartialEq)]
+#[cfg_attr(feature = "eders", derive(Serialize, Deserialize))]
+pub struct Cuboid<S> {
+    /// Dimensions of the box
+    pub dim: Vector3<S>,
+    half_dim: Vector3<S>,
+    corners: Vec<Point3<S>>,
+}
+
+impl<S> Cuboid<S>
+where
+    S: BaseFloat,
+{
+    /// Create a new rectangle primitive from component dimensions
+    pub fn new(dim_x: S, dim_y: S, dim_z: S) -> Self {
+        Self::new_impl(Vector3::new(dim_x, dim_y, dim_z))
+    }
+
+    /// Create a new rectangle primitive from a vector of component dimensions
+    pub fn new_impl(dim: Vector3<S>) -> Self {
+        Self {
+            dim,
+            half_dim: dim / (S::one() + S::one()),
+            corners: Self::generate_corners(&dim),
+        }
+    }
+
+    fn generate_corners(dimensions: &Vector3<S>) -> Vec<Point3<S>> {
+        let two = S::one() + S::one();
+        vec![
+            Point3::new(dimensions.x, dimensions.y, dimensions.z) / two,
+            Point3::new(-dimensions.x, dimensions.y, dimensions.z) / two,
+            Point3::new(-dimensions.x, -dimensions.y, dimensions.z) / two,
+            Point3::new(dimensions.x, -dimensions.y, dimensions.z) / two,
+            Point3::new(dimensions.x, dimensions.y, -dimensions.z) / two,
+            Point3::new(-dimensions.x, dimensions.y, -dimensions.z) / two,
+            Point3::new(-dimensions.x, -dimensions.y, -dimensions.z) / two,
+            Point3::new(dimensions.x, -dimensions.y, -dimensions.z) / two,
+        ]
+    }
+}
+
+impl<S> SupportFunction for Cuboid<S>
+where
+    S: BaseFloat,
+{
+    type Point = Point3<S>;
+
+    fn support_point<T>(&self, direction: &Vector3<S>, transform: &T) -> Point3<S>
+    where
+        T: Transform<Point3<S>>,
+    {
+        ::primitive::util::get_max_point(&self.corners, direction, transform)
+    }
+}
+
+impl<S> HasAabb for Cuboid<S>
+where
+    S: BaseFloat,
+{
+    type Aabb = Aabb3<S>;
+
+    fn get_bound(&self) -> Aabb3<S> {
+        Aabb3::new(
+            Point3::from_vec(-self.half_dim),
+            Point3::from_vec(self.half_dim),
+        )
+    }
+}
+
+impl<S> Discrete<Ray3<S>> for Cuboid<S>
+where
+    S: BaseFloat,
+{
+    fn intersects(&self, ray: &Ray3<S>) -> bool {
+        let min = Point3::new(-self.half_dim.x, -self.half_dim.y, -self.half_dim.z);
```

Comment:
could call `get_bound` here

Reply:
Yeah, just felt that the extra creation of an aabb wasn't necessary

Reply:
Then again, if I did that I can use the Aabb vs ray intersection tests for it, so yeah, changing it.